### PR TITLE
TensorIterator cuda launch configs update

### PR DESCRIPTION
**Pull request by jjsjann123**
Summary:
  Update launch configs for TensorIterator gpu_reduce_kernel. Enable flexible
block dimension to improve efficiency for reduction cases with small fast
dimension.

Previously TensorIterator launches blocks with fixed 32x16 threads.
For cases like:

  import torch
  torch.randn(2**20, 4, device='cuda').sum(0)

The fixed launch config does handle coalesced memory access efficiently.

Updated launch configure enables flexible block dimension. Combining with
improved reduction scheme (using flexible vertical / horizontal reduction
instead of limited warp / block reduction in the old code), it ensures optimal
memory access pattern even with reduction on dimension with small stride.

Possible future improvements:
1. Precise dynamic shared memory allocation.
2. Using warp shuffle for vertical (block_y) reduction.

### CODE REVIEW

**Comment by jjsjann123** [comment]
This is my monkey benchmarking. Notice the perf gain for small fast changing dimensions.

perf

**Comment by jjsjann123** [comment]
For visibility @ngimel @umanwizard @colesbury

**Comment by umanwizard** [comment]
We are getting this error on some internal builds:
```
aten/src/ATen/native/cuda/Reduce.cuh:34:12: error: right shift count >= width of type [-Werror=shift-count-overflow]
   n |= (n >> 32);
```

**Review by umanwizard — DISMISSED** [review]
see above comment

**Comment by jjsjann123** [comment]
My bad.
I thought I updated that two commits ago when I was changing the data type from int64 to int. Guess it didn't happen.

**Comment by apaszke** [comment]
I'm somewhat worried about the slowdowns. Why are we getting rid of warp reductions? Aren't those supposed to be fast?

**Comment by jjsjann123** [comment]
I'm not getting rid of warp reduction. Keeping them as necessary for the old launch config where block dimension as 32x16: https://github.com/pytorch/pytorch/pull/16224/files#diff-662693ef7b7f32fa32d7179b6614fc16R379

Renaming warp_reduce to block_x_reduce because as we have flexible block dimension, it might require a hybrid of shared memory reduction and warp reduction.

File: aten/src/ATen/native/cuda/Reduce.cuh [hunk 1]
```diff
@@ -331,22 +356,38 @@ struct ReduceOp {
...
 
-  C10_DEVICE arg_t warp_reduce(arg_t value) const {
-    for (int offset = 1; offset < warpSize; offset <<= 1) {
+  C10_DEVICE arg_t block_x_reduce(arg_t value, char* shared_memory) const {
+    int dim_x = blockDim.x;
+    arg_t* shared = (arg_t*)shared_memory;
+    if (dim_x > warpSize) {
+      int address_base = threadIdx.x + threadIdx.y*blockDim.x;
+      shared[address_base] = value;
+      for (int offset = dim_x/2; offset >= 32; offset >>= 1) {
```
Comment (iotamudelta) [inline aten/src/ATen/native/cuda/Reduce.cuh]: @umanwizard the hard-coded 32 here will not work for architectures that happen to not have a warp size of 32.

Comment (jjsjann123) [reply aten/src/ATen/native/cuda/Reduce.cuh]: Oops, good catch.
I'll update it with `warpSize`

**Comment by jjsjann123** [comment]
Test failure doesn't seem to be relevant. Merging ToT to see if it goes away

**Comment by soumith** [comment]
just fyi @jjsjann123 this is being reverted, we are seeing "illegal memory exception" and `RuntimeError: Creating MTGP constants failed. at caffe2/aten/src/THC/THCTensorRandom.cu:33 (when using torch.manual_seed)` after this PR was landed in hard-to-repro internal workloads.

**Comment by jjsjann123** [comment]
I think I figured out the issue. It's me overlooking the old heuristics for inter-block vs inter-warp reduction. Surprised this doesn't get caught by CI tests.

Working on a fix. Will also add tests that covers it.

**Pull request closed without merging**